test/javascripts/turbolinks_test.coffee

```diff
@@ -1,210 +1,946 @@
+          expectedScriptOrder.forEach (script, index) ->
+            assert.equal(script, actualExecutionOrder[index])
+          done()
+
+      it 'works in order BAC', (done) ->
+        window.actualExecutionOrder = []
+        visit url: 'threeScriptsInHeadBAC', ->
+          expectedScriptOrder = ['b', 'a', 'c']
+          expectedScriptOrder.forEach (script, index) ->
+            assert.equal(script, actualExecutionOrder[index])
+          done()
+
+      it 'works in order BCA', (done) ->
+        window.actualExecutionOrder = []
+        visit url: 'threeScriptsInHeadBCA', ->
+          expectedScriptOrder = ['b', 'c', 'a']
+          expectedScriptOrder.forEach (script, index) ->
+            assert.equal(script, actualExecutionOrder[index])
+          done()
+
+      it 'works in order CAB', (done) ->
+        window.actualExecutionOrder = []
+        visit url: 'threeScriptsInHeadCAB', ->
+          expectedScriptOrder = ['c', 'a', 'b']
+          expectedScriptOrder.forEach (script, index) ->
+            assert.equal(script, actualExecutionOrder[index])
+          done()
+
+      it 'works in order CBA', (done) ->
+        window.actualExecutionOrder = []
+        visit url: 'threeScriptsInHeadCBA', ->
+          expectedScriptOrder = ['c', 'b', 'a']
+          expectedScriptOrder.forEach (script, index) ->
+            assert.equal(script, actualExecutionOrder[index])
+          done()
+
+    it 'executes new scripts in the relative order they are present in the dom of the upstream document', (done) ->
+      window.actualExecutionOrder = []
+      visit url: 'secondLibraryOnly', ->
+        assert.equal(actualExecutionOrder[0], 'b')
+        visit url: 'threeScriptsInHeadABC', ->
+          assert.equal(actualExecutionOrder[1], 'a')
+          assert.equal(actualExecutionOrder[2], 'c')
+          done()
+
+    it 'performs a full page refresh when encountering a script with a new src that matches an existing data-turbolinks-track-script-as', (done) ->
+      visit url: 'singleScriptInHead', ->
+        visit url: 'singleScriptInHeadWithDifferentSourceButSameName', ->
+          assert(Turbolinks.fullPageNavigate.called, 'Should perform a full page refresh.')
+          done()
+
+    it 'does not perform a full page refresh when encountering a script with a new src that does not match a data-turbolinks-track-script-as', (done) ->
+      visit url: 'singleScriptInHead', ->
+        visit url: 'differentSingleScriptInHead', ->
+          assert(Turbolinks.fullPageNavigate.notCalled, 'Should not perform a full page refresh.')
+          done()
+
+  describe 'with partial page replacement', ->
+    it 'script tags are evaluated when they are the subject of a partial replace', (done) ->
+      window.globalStub = stub()
+      visit url: 'inlineScriptInBody', options: {partialReplace: true, onlyKeys: ['turbo-area']}, ->
+        assert(globalStub.calledOnce, 'Script tag was not evaluated :(')
+        done()
+
```

Comment:
That was basically the plan, but there was disagreement on whether we:
- include the polyfill in the project
- state that a polyfill is a dependency to use it in an older browser (and possibly break some of our/other's projects using it?)
- avoid them in favour of a callback solution (what we went with for now)